### PR TITLE
Add multisig Python SDK tooling, tutorial

### PR DESCRIPTION
**Pull request by alnoki**
This first commit adds multisig Python SDK tooling and a partial tutorial. It could use some initial review before additional content is added.

@clay-aptos is this layout/indexing within the doc site appropriate?

@davidiw are these class preliminary structures appropriate? Additionally, how do we go about using the [multisig authenticator](http://aptos.dev/guides/creating-a-signed-transaction#multisignature-transactions) data structure for transaction submission?

### CODE REVIEW

**Review by davidiw — COMMENTED** [review]
this is great! 

could we add some sort of validated encoding to either authenticator.py or to transaction.py for a proper end-to-end of into and out of transactions? we already have some for other types.

you can generate it even with the sdk, but I want something that is intended to be static here.

I'll need to review it more carefully in a bit

could you elaborate on what you mean "use it for transaction submission"?

**Review by clay-aptos — APPROVED** [review]
Thank you so much, Al! This LGTM with some edits I made directly to get links working in GitHub too. Yes, the navigation looks great.

Look out for your name listed as an author on this page (at the bottom) within a week of the PR being merged in. I will also announce in See what's new. Thanks again!

**Comment by alnoki** [comment]
> could we add some sort of validated encoding to either authenticator.py or to transaction.py for a proper end-to-end of into and out of transactions? we already have some for other types.
> 
> you can generate it even with the sdk, but I want something that is intended to be static here.
> 
> I'll need to review it more carefully in a bit

@davidiw I am not exactly sure what you are proposing with regards to validated encoding, but I am certainly open to more additions to the class APIs. I only implemented what was necessary to make some headway on the tutorial, and certainly skipped over some of the class APIs that were available on other types. But now I've got more eyes on it and welcome any additions that make things consistent.

> could you elaborate on what you mean "use it for transaction submission"?

I understand from the [`Authenticator` docs](https://aptos.dev/guides/creating-a-signed-transaction/#multisignature-transactions) that a multisig authenticator requires all of the individual signatures for the raw transaction, as well as a 4-byte big-endian bitmap of included signatures. What is not clear, however, is how to use these data structures to create a signature for the entire multi-ed25519 account. For example, here's an algorithm that is (probably) wrong, but is at least something that could be implemented in Python:

1. Concatenate the signatures with the bitmap
2. Hash the result
3. Use the hash as a quasi-private key to sign the raw transaction
4. The result is the multi-ed25519 signature

What is the correct way to generate a multi-ed25519 signature once the individual signatures and bitmap have been assembled?

**Comment by alnoki** [comment]
> Thank you so much, Al! This LGTM with some edits I made directly to get links working in GitHub too. Yes, the navigation looks great.
> 
> Look out for your name listed as an author on this page (at the bottom) within a week of the PR being merged in. I will also announce in See what's new. Thanks again!

@clay-aptos Sounds good, thanks for the edits. Feel free to update moving forward for docs consistency.

**Comment by alnoki** [comment]
@davidiw The latest commit adds multisig authenticator support, with successful serialization for BCS transaction submission (the tutorial now demonstrates transferring `APT` from the multisig account to one of the signatories).

Next planned installments for the tutorial:

1. Establish a mutisig account under a vanity address, via key rotation.
2. Publish a Move package from a multisig account.
3. Invoke a Move script from a multisig account.
4. Upgrade a Move package.

**Review by davidiw — COMMENTED** [review]
Something like this: https://github.com/aptos-labs/aptos-core/blob/f8283fbd85472b3f880a2457f4a794761a6eeb55/ecosystem/python/sdk/aptos_sdk/transactions.py#L582-L583

just more testing -- I don't like relying on e2e interaction to validate.

btw are you looking to get this in now or should we build out the entirety and review it in one fell swoop?

**Comment by alnoki** [comment]
> Something like this:
> 
> https://github.com/aptos-labs/aptos-core/blob/f8283fbd85472b3f880a2457f4a794761a6eeb55/ecosystem/python/sdk/aptos_sdk/transactions.py#L582-L583
> 
> just more testing -- I don't like relying on e2e interaction to validate.

@davidiw I was just trying to hit the practical e2e milestone, but if there is a sample of multisig input/output hexstrings I'd be glad to incorporate them for more formal testing. Is there an archive or similar of "truth sources" in this regard?

> btw are you looking to get this in now or should we build out the entirety and review it in one fell swoop?

I am leaning toward building out the entirety. I actually just started a little mini-review of a `RotationProofChallenge` that is not ready yet, so at the very least that needs to be tweaked first. Though I think the more comprehensive tutorial with all of the installments I mentioned above (package publication/upgrade as well as script invocation) will be ultimately more suitable for landing.

**Comment by alnoki** [comment]
> @davidiw The latest commit adds multisig authenticator support, with successful serialization for BCS transaction submission (the tutorial now demonstrates transferring `APT` from the multisig account to one of the signatories).
> 
> Next planned installments for the tutorial:
> 
> 1. Establish a mutisig account under a vanity address, via key rotation.
> 2. Publish a Move package from a multisig account.
> 3. Invoke a Move script from a multisig account.
> 4. Upgrade a Move package.

The most recent commit accomplishes additional milestone 1.

File: ecosystem/python/sdk/examples/multisig.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    bob_balance = rest_client.account_balance(bob.address())
+    print(f"Bob's balance:    {bob_balance}")
+    chad_balance = rest_client.account_balance(chad.address())
+    print(f"Chad's balance:   {chad_balance}")
+    multisig_balance = rest_client.account_balance(multisig_address)
+    print(f"Multisig balance: {multisig_balance}") # <:!:section_6
+
+    if wait_for_user: input("\nPress Enter to continue...")
+
+    # :!:>section_7
```
Comment (alnoki) [inline ecosystem/python/sdk/examples/multisig.py:147]: @jackyWYX here is an example of vanity address key rotation

**Comment by alnoki** [comment]
> 

`

> > @davidiw The latest commit adds multisig authenticator support, with successful serialization for BCS transaction submission (the tutorial now demonstrates transferring `APT` from the multisig account to one of the signatories).
> > Next planned installments for the tutorial:
> > 
> > 1. Establish a mutisig account under a vanity address, via key rotation.
> > 2. Publish a Move package from a multisig account.
> > 3. Invoke a Move script from a multisig account.
> > 4. Upgrade a Move package.
> 
> The most recent commit accomplishes additional milestone 1.

The most recent commit accomplishes additional milestone 2.

**Comment by alnoki** [comment]
@clay-aptos All of the planned content is now included in the tutorial, so it looks like the docs site portion is ready for you to take a final look to check if everything conforms with the stylistic standards from other tutorials.

**Comment by alnoki** [comment]
@davidiw I've included everything I had scoped out for this tutorial so it is ready to land on my end. As for the raw/signed hexstring verification you alluded to [above](https://github.com/aptos-labs/aptos-core/pull/6356#issuecomment-1407562251), that can easily be added later so long as there is a truth reference that is generated outside of the Python SDK.

When you are ready to land this in, the only thing left is that the [PyPI](https://pypi.org/search/?q=aptos) will need to be updated since I added new Python APIs that are called in the tutorial.

**Review by clay-aptos — APPROVED** [review]
Thanks for the updates, Al. I made a few small edits.

Once again, this is a fantastic addition I will gladly announce in See what's new and also add to this forming list of community contributions:
https://github.com/aptos-labs/aptos-core/pull/6417

**Comment by alnoki** [comment]
> Thanks for the updates, Al. I made a few small edits.
> 
> Once again, this is a fantastic addition I will gladly announce in See what's new and also add to this forming list of community contributions: #6417

@clay-aptos Sounds good to me, and thanks for looking it over!

**Review by davidiw — APPROVED** [review]
thanks!

**Comment by davidiw** [comment]
You need to run the appropriate formatters

docs: prettier '**/*.(tsx|ts|js)'
python: make fmt

**Comment by davidiw** [comment]
general
pre-commit run --all-files

**Comment by alnoki** [comment]
> You need to run the appropriate formatters
> 
> docs: prettier '**/*.(tsx|ts|js)' python: make fmt

> general pre-commit run --all-files

@davidiw d171ec8 applies formatters and lint checks pass.

Once this lands, the [PyPI](https://pypi.org/project/aptos-sdk/) just needs to be updated with the SDK additions.

**Comment by alnoki** [comment]
@davidiw @clay-aptos @JackyWYX @chen-robert @movekevin  the most recent commits merge in the Aptos Multisig Execution Expeditor (AMEE) tool with a corresponding tutorial.

This tooling is compatible with the current multisig schema and enables authentication key rotation, package publication, and script invocation from a multisig account, enabling arbitrary modifications to multisig accounts (adding/removing signatories, updating signature thresholds, etc.) and assorted governance workflows.

Notably, AMEE provides password protection for keyfiles so that they are not stored in plain text on disk, and invokes publication and script transactions using GitHub repo metadata. This enables transparency in the signing process, whereby signatories can actually look on GitHub and see what they are signing off on (see the tutorial in the docs site preview).

None of this is contingent on #5894 , which means that devs can engage in all of the above multisig activities using AMEE as soon as the merge lands: basically this is a working solution that allows for crucial multisig governance operations now, while the new multisig model is still not deployed on mainnet.

**Comment by gregnazario** [comment]
@movekevin can you do a quick double check of this and merge it in

**Comment by alnoki** [comment]
@davidiw would you prefer the PR be split up into one for just the Python code snippets section (steps 1 through 8 on the [docs site page](https://deploy-preview-6356--aptos-developer-docs.netlify.app/tutorials/your-first-multisig)), and another for the AMEE tool with tutorial (steps 9 on the [docs site page](https://deploy-preview-6356--aptos-developer-docs.netlify.app/tutorials/your-first-multisig))? The second portion is a tool that was developed while the first part was waiting to land, which should prove useful for other ecosystem developers using the current multi-ed25519 standard.

**Comment by davidiw** [comment]
> @davidiw would you prefer the PR be split up into one for just the Python code snippets section (steps 1 through 8 on the [docs site page](https://deploy-preview-6356--aptos-developer-docs.netlify.app/tutorials/your-first-multisig)), and another for the AMEE tool with tutorial (steps 9 on the [docs site page](https://deploy-preview-6356--aptos-developer-docs.netlify.app/tutorials/your-first-multisig))? The second portion is a tool that was developed while the first part was waiting to land, which should prove useful for other ecosystem developers using the current multi-ed25519 standard.

@alnoki , yeah if we could split this up, it would be great.

Might be useful to find some time to chat about the other components.

**Comment by alnoki** [comment]
> > @davidiw would you prefer the PR be split up into one for just the Python code snippets section (steps 1 through 8 on the [docs site page](https://deploy-preview-6356--aptos-developer-docs.netlify.app/tutorials/your-first-multisig)), and another for the AMEE tool with tutorial (steps 9 on the [docs site page](https://deploy-preview-6356--aptos-developer-docs.netlify.app/tutorials/your-first-multisig))? The second portion is a tool that was developed while the first part was waiting to land, which should prove useful for other ecosystem developers using the current multi-ed25519 standard.
> 
> @alnoki , yeah if we could split this up, it would be great.
> 
> Might be useful to find some time to chat about the other components.

@davidiw #6788 isolates the first batch of content per your request, and I'll be glad to chat more about the second batch once #6788 lands (note that the [PyPi](https://pypi.org/project/aptos-sdk/) will need to be updated to account for the version minor updates in the SDK modifications).

**Comment by movekevin** [comment]
The tutorial looks fine to me but let's separate out the SDK changes if possible.

**Comment by davidiw** [comment]
> The tutorial looks fine to me but let's separate out the SDK changes if possible.

I'm fine with that, I can work with @alnoki on the sdk side.

**Comment by davidiw** [comment]
> The tutorial looks fine to me but let's separate out the SDK changes if possible.

He did :) https://github.com/aptos-labs/aptos-core/pull/6788

**Comment by clay-aptos** [comment]
Hi @alnoki , I am noting we launched this:
https://aptos.dev/community/contributions/community-highlights

And would love to have this online. Please let me know how O can help. Thanks!

**Pull request closed without merging**